### PR TITLE
Wire up isChained filter | Closes #2648

### LINKED ISSUE #2648: Wire up isChained filter
Tickets filter state needs to be aware of the dates filter state to be able to use isChained

### PR DESCRIPTION
**Pull request by nicolad**
## Problem this Pull Request solves
- Adds filter to `useTicketsFilterBarService`

- when value for filter is `true`, tickets list should only show tickets that have been assigned to the dates that are visible in the dates list

- one solution might be to save a list of IDs for the currently visible dates into some state manager somewhere (filter state?other?) then use that to either:

     - requery apollo cache for tickets with relations to the list of date IDSs

     - filter the existing list of tickets using the date IDs and each tickets relations

### CODE REVIEW

File: assets/src/application/services/context/RelationsProvider.tsx [hunk 1]
```diff
@@ -1,17 +1,13 @@
...
 
 const RelationsContext = createContext<RelationsManager | null>(null);
 
 const { Provider, Consumer: RelationsConsumer } = RelationsContext;
 
-const RelationsProvider: React.FC<ProviderProps> = ({ children }) => {
+const RelationsProvider: React.FC = ({ children }) => {
 	const relations = useRelationsManager();
 	return <Provider value={relations}>{children}</Provider>;
 };
```
Comment (manzoorwanijk) [inline assets/src/application/services/context/RelationsProvider.tsx:13]: I have simplified the types for context providers

File: assets/src/application/ui/input/InlineEditInput/InlineEditInfinity.tsx [hunk 2]
```diff
@@ -9,6 +9,11 @@ const InlineEditInfinity: React.FC<TextProps> = ({ children, onChange, className
 	const [value, setValue] = useState(children);
 	const [isEditing, setIsEditing] = useState(false);
 
+	// if value updated by parent
+	useEffect(() => {
+		setValue(children);
+	}, [children]);
+
```
Comment (manzoorwanijk) [inline assets/src/application/ui/input/InlineEditInput/InlineEditInfinity.tsx:16]: There was a bug that ticket quantity did not reflect the change when changing the corresponding date capacity.

File: assets/src/application/services/context/RelationsProvider.tsx (hunk 1 above)
Comment (nicolad) [reply assets/src/application/services/context/RelationsProvider.tsx:13]: nice, this change adds more visibility :+1:

File: assets/src/application/ui/input/InlineEditInput/InlineEditInfinity.tsx (hunk 2 above)
Comment (nicolad) [reply assets/src/application/ui/input/InlineEditInput/InlineEditInfinity.tsx:16]: hmm, interesting

File: assets/src/domain/eventEditor/hooks/edtrState/reducer.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+const reducer: EdtrStateReducer = (state, action) => {
+	const { type, visibleDatetimeIds } = action;
+
+	switch (type) {
+		case 'SET_VISIBLE_DATETIME_IDS':
+			return { ...state, visibleDatetimeIds };
+
+		default:
+			throw new Error('Unexpected action');
```
Comment (nicolad) [inline assets/src/domain/eventEditor/hooks/edtrState/reducer.ts:11]: :+1: :smile:

File: assets/src/domain/eventEditor/services/filterState/datetimes/useDatesListFilterStateManager.ts [hunk 4]
```diff
@@ -55,7 +58,7 @@ const useDatesListFilterStateManager = (): FSM => {
 			setSales,
 			setStatus,
 		}),
-		[...Object.values(state), ...Object.values(entityFilterState.getState())]
+		[...datesFSValues, ...entityFSValues]
```
Comment (manzoorwanijk) [inline assets/src/domain/eventEditor/services/filterState/datetimes/useDatesListFilterStateManager.ts:61]: @nicolad You were not happy with this, so I changed it 😄

File: assets/src/domain/eventEditor/hooks/edtrState/useEdtrStateManager.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+	}, []);
+
+	return useMemo(
+		() => ({
+			...state,
+			getState,
+			setVisibleDatetimeIds,
+		}),
+		Object.values(state)
+	);
```
Comment (nicolad) [inline assets/src/domain/eventEditor/hooks/edtrState/useEdtrStateManager.ts:32]: nice approach, memo-first :+1:

File: assets/src/domain/eventEditor/services/filterState/datetimes/useDatesListFilterStateManager.ts (hunk 4 above)
Comment (nicolad) [inline assets/src/domain/eventEditor/services/filterState/datetimes/useDatesListFilterStateManager.ts:61]: :+1:

File: assets/src/domain/eventEditor/services/filterState/tickets/useIsChainedFilter.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+import { useRelations } from '@appServices/apollo/relations';
+import { Ticket } from '@edtrServices/apollo';
+import { useEdtrState } from '@edtrHooks/edtrState';
+import { entitiesWithGuIdInArray } from '@sharedServices/predicates';
+
+type IsChainedFilterCallback = (args: { isChained: boolean; tickets: Array<Ticket> }) => Array<Ticket>;
+
+type IsChainedFilterTuple = [IsChainedFilterCallback, any];
+
+const useIsChainedFilter = (): IsChainedFilterTuple => {
```
Comment (nicolad) [inline assets/src/domain/eventEditor/services/filterState/tickets/useIsChainedFilter.ts:13]: tuple way :+1:

File: assets/src/domain/eventEditor/services/filterState/tickets/useIsChainedFilter.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+	const visibleDatetimeIds = getState().visibleDatetimeIds;
+
+	const relatedTicketIds = keys(
+		pickBy((relations) => {
+			return (
+				hasPath(['datetimes'], relations) &&
+				pathOr([], ['datetimes'], relations).some((id) => visibleDatetimeIds.includes(id))
+			);
+		}, getData().tickets)
+	);
```
Comment (nicolad) [inline assets/src/domain/eventEditor/services/filterState/tickets/useIsChainedFilter.ts:26]: might be a good candidate for an extracted predicate which can refactored later, in another PR, a bit with `pipe` or `compose`

File: assets/src/domain/eventEditor/services/filterState/tickets/useIsChainedFilter.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+	const relatedTicketIdsStr = JSON.stringify(relatedTicketIds);
+
+	const callback = useCallback<IsChainedFilterCallback>(
+		({ isChained, tickets }) => {
+			// bail early
```
Comment (nicolad) [inline assets/src/domain/eventEditor/services/filterState/tickets/useIsChainedFilter.ts:31]: :)

File: assets/src/domain/eventEditor/services/filterState/tickets/useTicketsFilterBarService.ts [hunk 9]
```diff
@@ -18,6 +19,37 @@ const useTicketsFilterBarService = (): void => {
...
+	 * isChained filter needs special treatment :)
+	 *
+	 * Unlike other sorters and filters, isChained filter is dependent upon
+	 * external state which can change. Thus we need to update our filter callback
+	 * to make sure the stale state is not bound to the filter callback.
+	 */
+	const [isChainedFilter, isChainedDeps] = useIsChainedFilter();
+	// To avoid multiple filter registrations, we will store the aleady registered
+	// filter unsubscribe callback in ref to use it to remove the existing filter.
+	const chainedFilterRef = useRef<VoidFunction>();
```
Comment (nicolad) [inline assets/src/domain/eventEditor/services/filterState/tickets/useTicketsFilterBarService.ts:32]: interesting approach, maybe add a comment somewhere about the fact that mutating `ref` won't cause re-renders and this is used just as a global variable(kind of) which is not changed during renders.

Comment (manzoorwanijk) [reply assets/src/domain/eventEditor/services/filterState/tickets/useTicketsFilterBarService.ts:32]: That's understood about and is mentioned in its [docs](https://reactjs.org/docs/hooks-reference.html#useref)

File: assets/src/domain/eventEditor/services/filterState/datetimes/useDatesListFilterStateManager.ts (hunk 4 above)
Comment (nicolad) [reply assets/src/domain/eventEditor/services/filterState/datetimes/useDatesListFilterStateManager.ts:61]: yes, thank you, that was too crowded there :D 
need to keep variable distancing these times

File: assets/src/domain/eventEditor/services/context/EventContext/index.tsx [hunk 10]
```diff
@@ -1,14 +1,13 @@
...
-const EventEditorEventIdContext = createContext<number | null>(null);
+const EventIdContext = createContext<number | null>(null);
 
-const { Provider } = EventEditorEventIdContext;
+const { Provider } = EventIdContext;
 
-const EventEditorEventIdProvider: React.FC<ContextProviderProps> = ({ children }) => {
+const EventIdProvider: React.FC = ({ children }) => {
 	const toaster = useToaster();
 	const eventId = pathOr<number>(0, ['eeEditorData', 'event', 'dbId'], window);
```
Comment (manzoorwanijk) [inline assets/src/domain/eventEditor/services/context/EventContext/index.tsx:13]: The name for this context was too long and too decriptive 😸

File: assets/src/domain/eventEditor/services/context/EntityListContext/TicketsListContext.tsx [hunk 11]
```diff
@@ -11,15 +11,15 @@ export type TicketsListContextProps = EntityListContextProps<TicketsFilterStateM
 
 export const TicketsListContext = createContext<TicketsListContextProps>(null);
 
-export const TicketsListProvider: React.FC<ContextProviderProps> = ({ children }) => {
-	const datetimes = useTickets();
+export const TicketsListProvider: React.FC = ({ children }) => {
+	const tickets = useTickets();
```
[the remark this answers is not in the document]
Comment (manzoorwanijk) [reply assets/src/domain/eventEditor/services/context/EntityListContext/TicketsListContext.tsx:15]: yes 😄

**Pull request merged**